crud/src/main/java/com/redhat/lightblue/assoc/ep/Searches.java

```diff
@@ -0,0 +1,197 @@
+        for(ChildSlot slot:slots) {
+            associateDocs(parentDoc,slot,childDocs,aq);
+        }
+    }
+
+
+    /**
+     * Associate child documents with their parents. The association
+     * query is for the association from the child to the parent, so
+     * caller must flip it before sending it in if necessary. The
+     * caller also make sure parentDocs is a unique stream.
+     *
+     * @param parentDoc The parent document
+     * @param parentSlot The slot in parent docuemnt to which the results will be attached
+     * @param childDocs The child documents
+     * @param aq The association query from parent to child. This may
+     * not be the same association query between the blocks. If the
+     * child block is before the parent block, a new aq must be
+     * constructed for the association from the parent to the child
+     */
+    public static void associateDocs(ResultDocument parentDoc,
+                                     ChildSlot parentSlot,
+                                     List<ResultDocument> childDocs,
+                                     AssociationQuery aq) {
+        if(!childDocs.isEmpty()) {
+            LOGGER.debug("Associating docs");
+            ExecutionBlock childBlock=childDocs.get(0).getBlock();
+            ExecutionBlock parentBlock=parentDoc.getBlock();            
+            ArrayNode destNode=(ArrayNode)parentDoc.getDoc().get(parentSlot.getSlotFieldName()); 
+            BindQuery binders=parentDoc.getBindersForSlot(parentSlot,aq);
+            // No binders means all child docs will be added to the parent            
+            if(binders.getBindings().isEmpty()) {
+                if(destNode==null) {
+                    destNode=JsonNodeFactory.instance.arrayNode();
+                    parentDoc.getDoc().modify(parentSlot.getSlotFieldName(),destNode,true);
+                }
+                for(ResultDocument d:childDocs)
+                    destNode.add(d.getDoc().getRoot());
+            } else {
+                QueryExpression boundQuery=binders.iterate(aq.getQuery());
+                LOGGER.debug("Association query:{}",boundQuery);
+                QueryEvaluator qeval=QueryEvaluator.getInstance(boundQuery,childBlock.getMetadata());
+                for(ResultDocument childDoc:childDocs) {
+                    if(qeval.evaluate(childDoc.getDoc()).getResult()) {
+                        if(destNode==null) {
+                            destNode=JsonNodeFactory.instance.arrayNode();
+                            parentDoc.getDoc().modify(parentSlot.getSlotFieldName(),destNode,true);
+                        }
+                        destNode.add(childDoc.getDoc().getRoot());
+                    }
+                }
+            }
+        }
+    }
+
+
+    /**
+     * Combines queries with AND. Queries can be null, but at least one of them must be non-null
+     */
+    public static QueryExpression _and(QueryExpression... q) {
```

Comment:
Done.

On Tue, Mar 29, 2016 at 4:56 PM, Brandon Vulaj <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/assoc/ep/Searches.java
> <https://github.com/lightblue-platform/lightblue-core/pull/608#discussion_r57814788>
> :
>
> > +                        if(destNode==null) {
> > +                            destNode=JsonNodeFactory.instance.arrayNode();
> > +                            parentDoc.getDoc().modify(parentSlot.getSlotFieldName(),destNode,true);
> > +                        }
> > +                        destNode.add(childDoc.getDoc().getRoot());
> > +                    }
> > +                }
> > +            }
> > +        }
> > +    }
> > +
> > +
> > +    /**
> > +     * Combines queries with AND. Queries can be null, but at least one of them must be non-null
> > +     */
> > +    public static QueryExpression _and(QueryExpression... q) {
>
> Can we ditch the underscore?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/608/files/d72d06258b9b58a53294c8eb4a994145df771c4c#r57814788>
>